polkadot/node/core/pvf/benches/host_prepare_rococo_runtime.rs

```diff
@@ -0,0 +1,134 @@
+
+const TEST_PREPARATION_TIMEOUT: Duration = Duration::from_secs(30);
+
+struct TestHost {
+	#[allow(unused)]
+	cache_dir: tempfile::TempDir,
+	host: Mutex<ValidationHost>,
+}
+
+impl TestHost {
+	fn new_with_config<F>(handle: &Handle, f: F) -> Self
+	where
+		F: FnOnce(&mut Config),
+	{
+		let (prepare_worker_path, execute_worker_path) = testing::get_and_check_worker_paths();
+
+		let cache_dir = tempfile::tempdir().unwrap();
+		let mut config = Config::new(
+			cache_dir.path().to_owned(),
+			None,
+			prepare_worker_path,
+			execute_worker_path,
+		);
+		f(&mut config);
+		let (host, task) = start(config, Metrics::default());
+		let _ = handle.spawn(task);
+		Self { cache_dir, host: Mutex::new(host) }
+	}
+
+	async fn precheck_pvf(
+		&self,
+		code: &[u8],
+		executor_params: ExecutorParams,
+	) -> Result<PrepareStats, PrepareError> {
+		let (result_tx, result_rx) = futures::channel::oneshot::channel();
+
+		let code = sp_maybe_compressed_blob::decompress(code, 16 * 1024 * 1024)
+			.expect("Compression works");
+
+		self.host
+			.lock()
+			.await
+			.precheck_pvf(
+				PvfPrepData::from_code(
+					code.into(),
+					executor_params,
+					TEST_PREPARATION_TIMEOUT,
+					PrepareJobKind::Prechecking,
+				),
+				result_tx,
+			)
+			.await
+			.unwrap();
+		result_rx.await.unwrap()
+	}
+
+	async fn prune_all_artifacts(&self) -> Result<usize, std::io::Error> {
+		let (result_tx, result_rx) = futures::channel::oneshot::channel();
+
+		self.host.lock().await.prune_all_artifacts(result_tx).await.unwrap();
+		result_rx.await.unwrap()
+	}
+}
+
+fn host_prepare_rococo_runtime(c: &mut Criterion) {
+	polkadot_node_core_pvf_common::sp_tracing::try_init_simple();
+
+	let rt = tokio::runtime::Runtime::new().unwrap();
+
+	let host = TestHost::new_with_config(rt.handle(), |cfg| {
+		cfg.prepare_workers_hard_max_num = 1;
+	});
+
+	let blob = rococo_runtime::WASM_BINARY.unwrap();
+	let pvf = match sp_maybe_compressed_blob::decompress(&blob, 64 * 1024 * 1024) {
+		Ok(code) => PvfPrepData::from_code(
+			code.into_owned(),
+			ExecutorParams::default(),
+			Duration::from_secs(360),
+			PrepareJobKind::Compilation,
+		),
+		Err(e) => {
+			panic!("Cannot decompress blob: {:?}", e);
+		},
+	};
+
+	let mut group = c.benchmark_group("rococo");
+	group.sampling_mode(SamplingMode::Flat);
+	group.sample_size(20);
+	group.measurement_time(Duration::from_secs(240));
+	// The host spinning up a worker will be done in criterion's "warmup" stage, and not counted in
+	// the results.
+	group.bench_function("host: prepare Rococo runtime", |b| {
+		b.to_async(&rt).iter(|| async {
+			// `PvfPrepData` is designed to be cheap to clone, so cloning shouldn't affect the
+			// benchmark accuracy
+			let _stats = host.precheck_pvf(&pvf.clone().code(), Default::default()).await.unwrap();
+
+			// Delete the prepared artifact. Otherwise the next iterations will immediately finish.
+			let num_deleted = host.prune_all_artifacts().await.unwrap();
```

Comment:
If you just use https://docs.rs/criterion/latest/criterion/struct.AsyncBencher.html#method.iter_batched you don't need to add this hack that isn't required in the production code.

Or you just delete the artifacts manually.

Reply:
Thanks! It's not production code because of the `test-utils` gate, and deleting the artifacts doesn't work as explained above. `iter_batched` looks like a better solution though. 👍